Initial Google Implementation

josephlewis42 opened the pull request:
We've wrapped up our initial implementation and it's been approved by Google for public release. Thanks @acchen97 for getting the repo set up for us.

This initial release was built on work @erjohnso did that I ported over to using our Java clients after the other plugins started experiencing SSL issues with their ruby versions.

erjohnso commented:
Great work @josephlewis42 - this LGTM

robbavey commented on spec/outputs/google_pubsub_spec.rb:
```diff
@@ -0,0 +1,86 @@
+    end
+
+    it 'sends attributes' do
+      allow(pubsub_client).to receive(:publish_message)
```
Nit: This allow statement is unnecessary

robbavey commented on spec/outputs/google_pubsub_spec.rb:
```diff
@@ -0,0 +1,86 @@
+
+  describe '#stop' do
+    it 'calls shutdown on the pubsub client' do
+      allow(pubsub_client).to receive(:shutdown)
```
Nit: This allow statement is unnecessary

robbavey commented on spec/outputs/google_pubsub_spec.rb:
```diff
@@ -0,0 +1,86 @@
+
+  describe '#receive' do
+    it 'sends the message as JSON text' do
+      allow(pubsub_client).to receive(:publish_message)
```
Nit: This allow statement is unnecessary

robbavey commented on spec/outputs/pubsub/client_spec.rb:
```diff
@@ -0,0 +1,121 @@
+
+  describe '#publish_message' do
+    it 'builds a message with passed in attributes' do
+      allow(com.google.api.core.ApiFutures).to receive(:addCallback)
```
Nit: `allow` statement used in all these tests can be put in a `before` block inside `describe`

robbavey commented on lib/logstash/outputs/google_pubsub.rb:
```diff
@@ -0,0 +1,145 @@
+
+  # Send the batch once this delay has passed, from the time the first message
+  # is queued. (> 0, default: 5)
+  config :delay_threshold_secs, validate: :number, required: true, default: 5
```
While having a default value here makes the config item optional, having `required: true` here may cause confusion with the comment above stating the field is optional

robbavey commented on lib/logstash/outputs/google_pubsub.rb:
```diff
@@ -0,0 +1,145 @@
+
+  # Once this many messages are queued, send all the messages in a single call,
+  # even if the delay threshold hasn't elapsed yet. (< 1000, default: 100)
+  config :message_count_threshold, validate: :number, required: true, default: 100
```
No need for `required: true`

robbavey commented on lib/logstash/outputs/google_pubsub.rb:
```diff
@@ -0,0 +1,145 @@
+  # This includes full message payload size, including any attributes set.
+  #
+  # Default: 1MB
+  config :request_byte_threshold, validate: :number, required: true, default: 1_000_000
```
Consider using `validate: :bytes` - this enables users to specify values such as `1MB` or `256KiB` in their configurations rather than numeric values.

No need for `required: true`.

robbavey commented on lib/logstash/outputs/google_pubsub.rb:
```diff
@@ -0,0 +1,145 @@
+
+  # Attributes to add to the message in key: value formats.
+  # Keys and values MUST be strings.
+  config :attributes, validate: :hash, required: true, default: {}
```
No need for `required: true`

robbavey commented on lib/logstash/outputs/google_pubsub.rb:
```diff
@@ -0,0 +1,145 @@
+
+  def receive(event)
+    message = event.to_json
+    @logger.info("Sending message #{message}")
```
Should this be `debug` rather than `info`?

robbavey commented on lib/logstash/outputs/google_pubsub.rb:
```diff
@@ -0,0 +1,145 @@
+      @pubsub.build_message '', @attributes
+    rescue TypeError => e
+      message ='Make sure the attributes are string:string pairs'
+      @logger.info(message, error: e, attributes: @attributes)
```
Should this be `warn` or `error` rather than `info`?

robbavey commented on lib/logstash/outputs/pubsub/client.rb:
```diff
@@ -0,0 +1,124 @@
+
+        def construct_headers
+          gem_name = 'logstash-output-google_pubsub'
+          gem_version = '1.0.0'
```
Does this need to be kept in sync with the gem version?

fkoclas commented:
Hi,

I can't get the plugin to initialize properly due to a type error: `TypeError: cannot convert instance of class org.jruby.RubyString to class com.google.pubsub.v1.TopicName`

If I remove the topic from the configuration it complains that `topic => # SETTING MISSING` as you would expect.

I tried with both Logstash 5.6.10 and 6.3.1 and get the same error.

Here is the complete stacktrace:
> [ERROR][logstash.agent           ] Pipeline aborted due to error {:exception=>#<TypeError: cannot convert instance of class org.jruby.RubyString to class com.google.pubsub.v1.TopicName>, :backtrace=>["/opt/logstash/vendor/local_gems/510e3a38/logstash-output-google_pubsub-1.0.0-java/lib/logstash/outputs/pubsub/client.rb:80:in `initialize_google_client'", "/opt/logstash/vendor/local_gems/510e3a38/logstash-output-google_pubsub-1.0.0-java/lib/logstash/outputs/pubsub/client.rb:30:in `initialize'", "/opt/logstash/vendor/local_gems/510e3a38/logstash-output-google_pubsub-1.0.0-java/lib/logstash/outputs/google_pubsub.rb:114:in `register'", "/opt/logstash/logstash-core/lib/logstash/output_delegator_strategies/shared.rb:9:in `register'", "/opt/logstash/logstash-core/lib/logstash/output_delegator.rb:43:in `register'", "/opt/logstash/logstash-core/lib/logstash/pipeline.rb:290:in `register_plugin'", "/opt/logstash/logstash-core/lib/logstash/pipeline.rb:301:in `register_plugins'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/logstash-core/lib/logstash/pipeline.rb:301:in `register_plugins'", "/opt/logstash/logstash-core/lib/logstash/pipeline.rb:310:in `start_workers'", "/opt/logstash/logstash-core/lib/logstash/pipeline.rb:235:in `run'", "/opt/logstash/logstash-core/lib/logstash/agent.rb:408:in `start_pipeline'"]}

josephlewis42 commented:
@robbavey thanks for the feedback!

I changed a few things:

 * Bumped the version of the Pub/Sub library to the most recent version.
 * Added initial docs with examples.
 * Cleaned up the tests.
 * Changed `receive` to `multi_receive_encoded` (and tests around it) because it's really not acceptable to release a new plugin without it

------------

@fkoclas would you mind posting your config file? I can't seem to reproduce it in my setup. Mine looks like this:

```
input { stdin { } }

output {
    google_pubsub {
        project_id => "[redacted]"

        topic => "ls-input-google-pubsub-test-topic"
        attributes => {
		    "attr_key" => "attr_value"
		    "attr_key2" => "attrval2"
	    }

        json_key_file => "/home/[redacted]/key.json"
    }
}
```

fkoclas commented:
@josephlewis42 pretty much the same config:

```
input {
  stdin {
  }
}

output {
  google_pubsub {
    project_id => "<snip>"
    topic => "elasticsearch_indexing"
    json_key_file => "/etc/secrets/serviceaccount.json"
  }
}
```

Interestingly I did some more testing, and it turns out that it works on the `openjdk:8-jre` docker image (debian-based), but not on the `openjdk:8-jre-alpine` docker image which is what I have been using to run logstash for nearly the past two years.

robbavey commented:
@karenzone Can you please review the documentation additions here?

karenzone commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,51 @@
+
+include::resources.asciidoc[]
+
+include::configuration.asciidoc[]
```
Please add content from `examples.asciidoc`, `resources.asciidoc`, and `configuration.asciidoc` into `docs/index.asciidoc` file rather than using include:: statements.  We use scripts to generate various outputs and versions, and a single asciidoc file works best.

karenzone commented on README.md:
```diff
@@ -1,2 +1,96 @@
+# Logstash Output Google Pubsub
+
+A LogStash plugin to upload log events to https://cloud.google.com/pubsub/[Google Cloud Pubsub].
+Events are batched and uploaded in the background for the sake of efficiency.
```
Please change instances of `LogStash` to `Logstash`

josephlewis42 commented:
@karenzone I merged those files and changed LogStash -> Logstash

@robbavey License headers are gone and updated the jars include file

Thank you both for taking the time to help review this! :-)

karenzone commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,231 @@
+:plugin: google_pubsub
+:type: output
+:default_codec: plain
+
```
The`lib/logstash/outputs/google_pubsub.rb` file says the default codec is `json`.  If that's true, then set it here. 
`:default_codec: json`

josephlewis42 replied:
Whoops, I knew I had to change it somewhere and added it at the bottom of the file instead :man_facepalming:

karenzone commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,231 @@
+Note: While this project is partially maintained by Google, this is not an official Google product.
+
+=== Environment Configuration
+
```
Add one more `=` to heading so content doesn't push to a new page. 
`==== Environment Configuration`

karenzone commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,231 @@
+
+
+====== Different Codecs
+
```
This heading exceeds the levels we can render.  Can you remove one `=` to set it to `===== Different Codecs`?
If this heading must be one level down from Attributes, we can try surrounding heading  with `*` to make it bold.

karenzone commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,231 @@
+[id="plugins-{type}s-{plugin}-common-options"]
+include::{include_path}/{type}.asciidoc[]
+
+:default_codec!: json
```
This setting clears the default codec value set at the topic of the file. 
Should be `:default_codec!:`

josephlewis42 commented:
@robbavey and @karenzone thank you so much for the reviews, I'll merge this in and push a 1.0.0 gem tonight when I get home.

@fkoclas I'll open an issue and mention you in it so we can dig into the specific problem with the Alpine combo.

karenzone commented:
@josephlewis42  Thanks for your contribution and for the kind words. It's been a pleasure working with you on this.  FYI, the plugin is in the Versioned Plugin doc now (https://www.elastic.co/guide/en/logstash-versioned-plugins/current/v1.0.0-plugins-outputs-google_pubsub.html).  I have a few more steps to complete, and then it will appear in the Logstash Reference. Thanks again!

karenzone commented:
Hi @josephlewis42.  I'll be pulling your doc into the Logstash Reference Guide shortly.  While I was testing, I noticed that the `https://cloud.google.com/pubsub/pricing-summary/[Cloud Pub/Sub Pricing]` link under `Additional Resources` is not resolving.  I can merge the PR to include the plugin, but thought you'd want to know about the broken link so that we can get it fixed.
